File: src/device/class/char.c

// SPDX-FileCopyrightText: 2025 Julian Scheffers <julian@scheffers.net>
// SPDX-FileType: SOURCE
// SPDX-License-Identifier: MIT

#include "device/class/char.h"

#include "badgeros-abi/errno.h"
#include "sched/sync/mutex.h"



// Read bytes from the device.
errno_size_t device_char_read(device_char_t *device, void __user *rdata, size_t rdata_len, bool nonblock) {
    mutex_lock_shared(&device->base.driver_mtx);
    if (!device->base.driver) {
        mutex_unlock_shared(&device->base.driver_mtx);
        return -ENOENT;
    }
    driver_char_t const *driver = (void *)device->base.driver;
    errno_size_t         res    = driver->read(device, rdata, rdata_len, nonblock);
    mutex_unlock_shared(&device->base.driver_mtx);
    return res;
}

// Write bytes to the device.
errno_size_t device_char_write(device_char_t *device, void const __user *wdata, size_t wdata_len, bool nonblock) {
    mutex_lock_shared(&device->base.driver_mtx);
    if (!device->base.driver) {
        mutex_unlock_shared(&device->base.driver_mtx);
        return -ENOENT;
    }
    driver_char_t const *driver = (void *)device->base.driver;
    errno_size_t         res    = driver->write(device, wdata, wdata_len, nonblock);
    mutex_unlock_shared(&device->base.driver_mtx);
    return res;
}

File: src/dev2/class/char.rs
// SPDX-FileCopyrightText: 2026 Julian Scheffers <julian@scheffers.net>
// SPDX-FileType: SOURCE
// SPDX-License-Identifier: MIT

use alloc::vec::Vec;

use crate::{
    bindings::error::EResult,
    dev2::Device,
    kernel::sync::waitlist::Waitlist,
    process::usercopy::{UserSlice, UserSliceMut},
};

/// A character device (e.g. 16550 UART, /dev/null, etc).
pub trait CharDevice: Device {
    /// Is this a TTY?
    fn is_tty(&self) -> bool {
        true
    }

    /// Get current polling status flags.
    fn poll(&self) -> u32;

    /// Collect waitlists for the requested poll interest flags.
    fn poll_waitlists<'a>(&'a self, interest: u32, collect: &mut Vec<&'a Waitlist>) -> EResult<()>;

    /// Read raw bytes from this device (ignoring termios).
    /// Upon (partial) success, returns how many bytes were read.
    fn read_raw(&self, rdata: UserSliceMut<u8>, nonblock: bool) -> EResult<usize>;

    /// Write raw bytes to this device (ignoring termios).
    /// Upon (partial) success, returns how many bytes were written.
    fn write_raw(&self, wdata: UserSlice<u8>, nonblock: bool) -> EResult<usize>;
}
